storage/statedb/hasher.go

```diff
@@ -183,34 +193,43 @@ func (h *hasher) hashChildren(original node, db *Database, onRoot bool) (node, n
 // store hashes the node n and if we have a storage layer specified, it writes
 // the key/value pair to it and tracks any node->child references as well as any
 // node->external trie references.
-func (h *hasher) store(n node, db *Database, force bool) (node, uint16) {
+func (h *hasher) store(n node, db *Database, force bool, onRoot bool) (node, uint16) {
 	// Don't store hashes or empty nodes.
 	if _, isHash := n.(hashNode); n == nil || isHash {
 		return n, 0
 	}
+	// hash is for the merkle proof. hash = Keccak(rlp.Encode(nodeForHashing(n)))
+	// lenEncoded is for Database size accounting. lenEncoded = len(rlp.Encode(nodeForStoring(n)))
 	hash, _ := n.cache()
 	lenEncoded := n.lenEncoded()
-	if hash == nil || lenEncoded == 0 {
-		// Generate the RLP encoding of the node
+
+	// Calculate lenEncoded if not set
+	if lenEncoded == 0 {
 		h.tmp.Reset()
-		if err := rlp.Encode(&h.tmp, n); err != nil {
+		if err := rlp.Encode(&h.tmp, h.nodeForStoring(n)); err != nil {
 			panic("encode error: " + err.Error())
 		}
-
 		lenEncoded = uint16(len(h.tmp))
 	}
 	if lenEncoded < 32 && !force {
 		return n, lenEncoded // Nodes smaller than 32 bytes are stored inside their parent
 	}
+
+	// Calculate hash if not set
 	if hash == nil {
-		hash = h.makeHashNode(h.tmp)
+		h.tmp.Reset()
+		if err := rlp.Encode(&h.tmp, h.nodeForHashing(n)); err != nil {
+			panic("encode error: " + err.Error())
+		}
```

Comment:
It's a question.. Can I ask why we don't need to update the lenEncoded and check the length value when hash is nil?

Reply:
Because lenEncoded is already set at lines 207 to 213
and the length < 32 check is done separately at line 214 to 216?

Reply:
Updated in 0ae4fd54. Now the logic should be the same as original.